File: 2024/10/solve.py
print("Starting Day 10 solver")
try:
    with open('input.txt') as f:
        data = f.read().splitlines()
except:
    print("Your input file must be named input.txt and placed in this directory")

tab = []
for line in data:
    tab.append(list(map(int, [e for e in line])))

def get_trail_head_score(x, y, visited=None):
    if visited is None:
        visited = []
    visited.append((x, y))
    if tab[x][y] == 9:
        return 1
    else:
        count = 0
        if x > 0  and tab[x-1][y] == tab[x][y]+1 and (x-1, y) not in visited:
            count += get_trail_head_score(x-1, y, visited)
        if x < len(tab)-1 and tab[x+1][y] == tab[x][y]+1 and (x+1, y) not in visited:
            count += get_trail_head_score(x+1, y, visited)
        if y > 0 and tab[x][y-1] == tab[x][y]+1 and (x, y-1) not in visited:
            count += get_trail_head_score(x, y-1, visited)
        if y < len(tab[0])-1 and tab[x][y+1] == tab[x][y]+1 and (x, y+1) not in visited:
           count += get_trail_head_score(x, y+1, visited)
        return count

print("\n------------------------------\nComputing first part")
count = 0
for x in range(len(tab)):
    for y in range(len(tab[0])):
        if tab[x][y] == 0:
            count += get_trail_head_score(x, y)
print(f"Answer : {count}\n------------------------------\n")

def get_trail_head_rating(x, y, visited=None):
    if visited is None:
        visited = []
    if tab[x][y] == 9:
        return 1
    else:
        count = 0
        if x > 0  and tab[x-1][y] == tab[x][y]+1 and (x-1, y) not in visited:
            count += get_trail_head_rating(x-1, y, visited + [(x, y)])
        if x < len(tab)-1 and tab[x+1][y] == tab[x][y]+1 and (x+1, y) not in visited:
            count += get_trail_head_rating(x+1, y, visited + [(x, y)])
        if y > 0 and tab[x][y-1] == tab[x][y]+1 and (x, y-1) not in visited:
            count += get_trail_head_rating(x, y-1, visited + [(x, y)])
        if y < len(tab[0])-1 and tab[x][y+1] == tab[x][y]+1 and (x, y+1) not in visited:
           count += get_trail_head_rating(x, y+1, visited + [(x, y)])
        return count

print("\n------------------------------\nComputing second part")
count = 0
for x in range(len(tab)):
    for y in range(len(tab[0])):
        if tab[x][y] == 0:
            count += get_trail_head_rating(x, y)
print(f"Answer : {count}\n------------------------------")

File: 2024/10/solve.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var tab [][]int

func main() {
	fmt.Println("Starting Day 10 solver")

	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Your input file must be named input.txt and placed in this directory")
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		row := []int{}
		for _, e := range strings.Split(line, "") {
			num, _ := strconv.Atoi(e)
			row = append(row, num)
		}
		tab = append(tab, row)
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	fmt.Println("\n------------------------------\nComputing first part")
	count := 0
	for x := range tab {
		for y := range tab[0] {
			if tab[x][y] == 0 {
				count += getTrailHeadScore(x, y, nil)
			}
		}
	}
	fmt.Printf("Answer : %d\n------------------------------\n", count)

	fmt.Println("\n------------------------------\nComputing second part")
	count = 0
	for x := range tab {
		for y := range tab[0] {
			if tab[x][y] == 0 {
				count += getTrailHeadRating(x, y, nil)
			}
		}
	}
	fmt.Printf("Answer : %d\n------------------------------\n", count)
}

func getTrailHeadScore(x, y int, visited [][2]int) int {
	if visited == nil {
		visited = [][2]int{}
	}
	visited = append(visited, [2]int{x, y})
	if tab[x][y] == 9 {
		return 1
	} else {
		count := 0
		if x > 0 && tab[x-1][y] == tab[x][y]+1 && !contains(visited, [2]int{x - 1, y}) {
			count += getTrailHeadScore(x-1, y, visited)
		}
		if x < len(tab)-1 && tab[x+1][y] == tab[x][y]+1 && !contains(visited, [2]int{x + 1, y}) {
			count += getTrailHeadScore(x+1, y, visited)
		}
		if y > 0 && tab[x][y-1] == tab[x][y]+1 && !contains(visited, [2]int{x, y - 1}) {
			count += getTrailHeadScore(x, y-1, visited)
		}
		if y < len(tab[0])-1 && tab[x][y+1] == tab[x][y]+1 && !contains(visited, [2]int{x, y + 1}) {
			count += getTrailHeadScore(x, y+1, visited)
		}
		return count
	}
}

func getTrailHeadRating(x, y int, visited [][2]int) int {
	if visited == nil {
		visited = [][2]int{}
	}
	if tab[x][y] == 9 {
		return 1
	} else {
		count := 0
		if x > 0 && tab[x-1][y] == tab[x][y]+1 && !contains(visited, [2]int{x - 1, y}) {
			count += getTrailHeadRating(x-1, y, append(visited, [2]int{x, y}))
		}
		if x < len(tab)-1 && tab[x+1][y] == tab[x][y]+1 && !contains(visited, [2]int{x + 1, y}) {
			count += getTrailHeadRating(x+1, y, append(visited, [2]int{x, y}))
		}
		if y > 0 && tab[x][y-1] == tab[x][y]+1 && !contains(visited, [2]int{x, y - 1}) {
			count += getTrailHeadRating(x, y-1, append(visited, [2]int{x, y}))
		}
		if y < len(tab[0])-1 && tab[x][y+1] == tab[x][y]+1 && !contains(visited, [2]int{x, y + 1}) {
			count += getTrailHeadRating(x, y+1, append(visited, [2]int{x, y}))
		}
		return count
	}
}

func contains(slice [][2]int, item [2]int) bool {
	for _, v := range slice {
		if v == item {
			return true
		}
	}
	return false
}
